lib/internal/url.js

```diff
@@ -542,125 +524,92 @@ ObjectDefineProperties(URLSearchParams.prototype, {
+                         hasOpaquePath) {
   const ctx = this[context];
-  ctx.flags = flags;
-  ctx.scheme = protocol;
-  ctx.username = (flags & URL_FLAGS_HAS_USERNAME) !== 0 ? username : '';
-  ctx.password = (flags & URL_FLAGS_HAS_PASSWORD) !== 0 ? password : '';
-  ctx.port = port;
-  ctx.path = (flags & URL_FLAGS_HAS_PATH) !== 0 ? path : [];
-  ctx.query = query;
-  ctx.fragment = fragment;
+  ctx.href = href;
+  ctx.origin = origin;
+  ctx.protocol = protocol;
   ctx.host = host;
+  ctx.hostname = hostname;
+  ctx.pathname = pathname;
+  ctx.search = search;
+  ctx.username = username;
+  ctx.password = password;
+  ctx.port = port;
+  ctx.hash = hash;
+  // TODO(@anonrig): Remove hasHost and hasOpaquePath when kFormat is removed.
+  ctx.hasHost = hasHost;
+  ctx.hasOpaquePath = hasOpaquePath;
   if (!this[searchParams]) { // Invoked from URL constructor
     this[searchParams] = new URLSearchParams();
     this[searchParams][context] = this;
   }
-  initSearchParams(this[searchParams], query);
-}
-
-function onParseError(input, flags) {
-  throw new ERR_INVALID_URL(input);
+  initSearchParams(this[searchParams], ctx.search);
 }
 
-function onParseProtocolComplete(flags, protocol, username, password,
-                                 host, port, path, query, fragment) {
-  const ctx = this[context];
-  if ((flags & URL_FLAGS_SPECIAL) !== 0) {
-    ctx.flags |= URL_FLAGS_SPECIAL;
-  } else {
-    ctx.flags &= ~URL_FLAGS_SPECIAL;
-  }
-  ctx.scheme = protocol;
-  ctx.port = port;
-}
-
-function onParseHostnameComplete(flags, protocol, username, password,
-                                 host, port, path, query, fragment) {
-  const ctx = this[context];
-  if ((flags & URL_FLAGS_HAS_HOST) !== 0) {
-    ctx.host = host;
-    ctx.flags |= URL_FLAGS_HAS_HOST;
-  } else {
-    ctx.host = null;
-    ctx.flags &= ~URL_FLAGS_HAS_HOST;
-  }
-}
-
-function onParsePortComplete(flags, protocol, username, password,
-                             host, port, path, query, fragment) {
-  this[context].port = port;
-}
-
-function onParseHostComplete(flags, protocol, username, password,
-                             host, port, path, query, fragment) {
-  ReflectApply(onParseHostnameComplete, this, arguments);
-  if (port !== null || ((flags & URL_FLAGS_IS_DEFAULT_SCHEME_PORT) !== 0))
-    ReflectApply(onParsePortComplete, this, arguments);
-}
-
-function onParsePathComplete(flags, protocol, username, password,
-                             host, port, path, query, fragment) {
-  const ctx = this[context];
-  if ((flags & URL_FLAGS_HAS_PATH) !== 0) {
-    ctx.path = path;
-    ctx.flags |= URL_FLAGS_HAS_PATH;
-  } else {
-    ctx.path = [];
-    ctx.flags &= ~URL_FLAGS_HAS_PATH;
-  }
-
-  // The C++ binding may set host to empty string.
-  if ((flags & URL_FLAGS_HAS_HOST) !== 0) {
-    ctx.host = host;
-    ctx.flags |= URL_FLAGS_HAS_HOST;
-  }
+function isURLThis(self) {
+  return (self !== undefined && self !== null && self[context] !== undefined);
```

Comment:
nit: ideally this should use `Object.prototype.hasOwnProperty` to avoid prototype lookup. (e.g., `Object.create(new URL(…))` should not be considered a URL)

Reply:
I'm adding a TODO for this. Thank you.